### PR TITLE
Allow clients to limit the amount of data that navigate to returns.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/services/services.ts [hunk 1]
```diff
@@ -4686,22 +4686,32 @@ module ts {
...
                     }
                 }
             });
 
+            rawItems.sort((i1, i2) => i1.matchKind - i2.matchKind);
+            if (maxResultCount !== undefined) {
+                rawItems = rawItems.slice(0, maxResultCount);
+            }
+
+            var items = map(rawItems, i => {
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: i -> item

File: src/services/services.ts [hunk 2]
```diff
@@ -4672,7 +4672,7 @@ module ts {
             // default NavigateTo approach: if search term contains only lower-case chars - use case-insensitive search, otherwise switch to case-sensitive version
             var searchTerms = map(terms, t => ({ caseSensitive: hasAnyUpperCaseCharacter(t), term: t }));
 
-            var items: NavigateToItem[] = [];
+            var rawItems: { name: string; fileName: string; matchKind: MatchKind; declaration: Declaration }[] = [];
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: I would really prefer a type alias here, like `RawNavigateToItem` or `NavigateToMatch`

File: src/services/services.ts [hunk 3]
```diff
@@ -4686,22 +4686,32 @@ module ts {
...
+                    kind: getNodeKind(declaration),
+                    kindModifiers: getNodeModifiers(declaration),
+                    matchKind: MatchKind[i.matchKind],
+                    fileName: i.fileName,
+                    textSpan: createTextSpanFromBounds(declaration.getStart(), declaration.getEnd()),
+                    // TODO(jfreeman): What should be the containerName when the container has a computed name?
+                    containerName: container && container.name ? (<Identifier>container.name).text : "",
+                    containerKind: container && container.name ? getNodeKind(container) : ""
+                };
+            });
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Create a helper function `makeNavigateToItemFromMatch`

File: src/services/services.ts [hunk 4]
```diff
@@ -1381,8 +1381,8 @@ module ts {
     enum MatchKind {
         none = 0,
         exact = 1,
-        substring = 2,
-        prefix = 3
+        prefix = 2,
+        substring = 3,
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Why are these in lowercase?

Comment (CyrusNajmabadi) [inline src/services/services.ts]: Because they're part of our public JSON Api.  Changing them would break our clients.

**Pull request merged**